src/main/java/com/google/sps/servlets/MapServlet.java

```diff
@@ -1,54 +1,81 @@
 package com.google.sps.servlets;
-
-import com.google.template.soy.SoyFileSet;
-import com.google.template.soy.tofu.SoyTofu;
+ 
+import static com.google.sps.utils.SoyRendererUtils.getOutputString;
+import static com.google.sps.utils.StringConstants.*;
+ 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.KeyFactory;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import com.google.sps.utils.ServletUtils;
 import java.io.File;
 import java.io.IOException;
+import java.util.HashSet;
 import java.util.Scanner;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
-
-
+ 
+ 
 @WebServlet("/map")
 public class MapServlet extends HttpServlet {
-
+ 
     /**
      * Adds user's groups as checkboxes to the map drawer
      * Allows users to select which groups they want displayed on the map
      */
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-
+        ServletUtils.enforceUserLogin(request, response);
+ 
         // Reads API Key and HTTP referer from file
         ClassLoader classLoader = MapServlet.class.getClassLoader();
-        // TODO change file name to the constant in Valeria's PR once merged
-        File apiKeyFile = new File(classLoader.getResource("keys.txt").getFile());
+        File apiKeyFile = new File(classLoader.getResource(KEYS_TXT_FILE).getFile());
         Scanner scanner = new Scanner(apiKeyFile);
         final String apiKey = scanner.nextLine();
         scanner.close();
 
-        // Hard coded user groups for testing, 
-        // Will get groups from user/datastore after functionality is verified
-        ImmutableMap<String, String> groupMap = ImmutableMap.of("groupName", "Black Lives Matter",
-            "groupID", "123");
-        ImmutableMap<String, String> groupMap2 = ImmutableMap.of("groupName", "Sierra Club",
-            "groupID", "456");
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        String userEmail = request.getUserPrincipal().getName();
+        // Stores the IDs for all the groups the user ahs joined
+        HashSet<Long> userGroups = new HashSet<Long>();
+
+        try {
+            Entity user = datastore.get(KeyFactory.createKey(USER_KIND, userEmail));
+            userGroups = (HashSet<Long>) user.getProperty(GROUPS_KEY);
+
+        } catch (EntityNotFoundException newUser) {
+            // if the user does not exist, create one and add to datastore
+            Entity user = new Entity(USER_KIND, userEmail);
```

Comment:
Is there a way to detect if a user does not exist without waiting for an exception to be thrown?

Exceptions should really be used for actual exceptions that you can't really predict. If you can predict it (which it looks like you can here), use an if/else and handle the edge case.